### ISSUE TITLE
inference problem with constraints involving conditionals and generics 

### ISSUE THREAD

**Issue by zpdDG4gta8XKpMCd** [microsoft/TypeScript#29662]
the following is seen in typescript@3.4.0-dev.20190131:

```ts
declare function onlyNullablePlease<T extends (null extends T ? any : never)>(value: T): void;

declare var z: string | null;
onlyNullablePlease(z); // works as expected

declare var y: string;
onlyNullablePlease(y); // error as expected

function f<T>(t: T) {
    var x: T | null = Math.random() > 0.5 ? null : t;
    onlyNullablePlease(x); // should work, unexpected error:
    /*
Argument of type 'T | null' is not assignable to parameter of type 'null extends T | null ? any : never'.
  Type 'null' is not assignable to type 'null extends T | null ? any : never'.ts(2345)
    */
}
```

**Comment by jack-williams** [comment]
I suspect this PR #29338 introduced the delta, not the one you link. I also think this is the same issue as #29505. Wrapping the check and extends types in a 1-tuple fixes is a short-term fix.

```ts
declare function onlyNullablePlease<T extends ([null] extends [T] ? any : never)>(value: T): void;
```

**Comment by aleksey-bykov** [comment]
fyi, the problem is a major breaking change in our project, it prevents us from upgrading to the latest version of typescript, the workaround by @jack-williams helped though

**Comment by weswigham** [comment]
Personal request: Use `unknown` instead of `any` - type parameters extending `any` are currently.... strange (they're kinda checked and kinda not checked). See #29571 where we plan to fix the strangeness.

**Comment by aleksey-bykov** [comment]
unknown doesn't help

**Comment by weswigham** [comment]
Oh, yeah, it doesn't help here, but I just wanted to inform that `extends any` behaves poorly and might want to be avoided, since `extends unknown` behaves far better.
